org.eclipse.lsp4e/src/org/eclipse/lsp4e/MarkerAttributeComputer.java

```diff
@@ -11,89 +11,47 @@
- * attributes directly to the map returned by
- * {@link #computeMarkerAttributes(IDocument, Diagnostic)}. An attribute
- * containing the language server ID is added later to this map.
- * <p>
  */
 public class MarkerAttributeComputer {
 
-	// Specific marker attributes defined by LSP4E
-	public static final String LSP_DIAGNOSTIC = "lspDiagnostic"; //$NON-NLS-1$
-
 	/**
-	 * Defines the expected number of attributes that will be computed, it is used
-	 * to optimize size of the map at creation time.
+	 * Defines the expected number of attributes that will be added during
+	 * customization, it is used to optimize size of the map at creation time.
 	 *
 	 * @return the expected number of attributes that will be computed
 	 */
 	public int attributeCount() {
-		return 8;
+		return 0;
 	}
 
 	/**
-	 * Computes the attributes of a marker for the given document and diagnostic.
+	 * Customizes the attributes of a marker for the given document, diagnostic and resource,
+	 * adding new attributes if necessary.
 	 *
-	 * @param document,
+	 * @param document
 	 *            the {@link Document} attached to the given resource
-	 * @param diagnostic,
-	 *            the {@link Diagnostic} to me mapped to a marker.
-	 * @return a map with the marker attributes
+	 * @param diagnostic
+	 *            the {@link Diagnostic} to me mapped to a marker
+	 * @param resource
+	 *            the {@link IResource} that contains the document
+	 * @param attributes
+	 *            a map with the attributes for the marker
 	 */
-	public Map<String, Object> computeMarkerAttributes(@Nullable IDocument document, @NonNull Diagnostic diagnostic) {
-		Map<String, Object> attributes = new HashMap<>(attributeCount());
-		attributes.put(MarkerAttributeComputer.LSP_DIAGNOSTIC, diagnostic);
-		attributes.put(IMarker.MESSAGE, diagnostic.getMessage());
-		attributes.put(IMarker.SEVERITY, LSPEclipseUtils.toEclipseMarkerSeverity(diagnostic.getSeverity()));
-
-		if (document != null) {
-			Range range = diagnostic.getRange();
-			int documentLength = document.getLength();
-			try {
-				int start = Math.min(LSPEclipseUtils.toOffset(range.getStart(), document), documentLength);
-				int end = Math.min(LSPEclipseUtils.toOffset(range.getEnd(), document), documentLength);
-				int lineOfStartOffset = document.getLineOfOffset(start);
-				if (start == end && documentLength > end) {
-					end++;
-					if (document.getLineOfOffset(end) != lineOfStartOffset) {
-						start--;
-						end--;
-					}
-				}
-				attributes.put(IMarker.CHAR_START, start);
-				attributes.put(IMarker.CHAR_END, end);
-				attributes.put(IMarker.LINE_NUMBER, lineOfStartOffset + 1);
-			} catch (BadLocationException ex) {
-				LanguageServerPlugin.logError(ex);
-			}
-		}
-		return attributes;
+	public void customizeMarkerForDiagnostic(@Nullable IDocument document, @NonNull Diagnostic diagnostic,
+			@NonNull IResource resource, @NonNull Map<String, Object> attributes) {
+		// empty

```

Comment:
What about making it an interface then, and attribute count be a default method?

Reply:
Yes, let us do that. We are back to the very first implementation with an interface, I am happy with that :)